src/feat/feature-functions.cc

```diff
@@ -26,30 +26,96 @@
 
 namespace kaldi {
 
-void ComputePowerSpectrum(VectorBase<BaseFloat> *waveform) {
-  int32 dim = waveform->Dim();
+template<class Real>
+void ComputePowerSpectrum(VectorBase<Real> *complex_fft) {
+  int32 dim = complex_fft->Dim();
+  int32 half_dim = dim/2;
+  Real first_energy = (*complex_fft)(0) * (*complex_fft)(0),
+      last_energy = (*complex_fft)(1) * (*complex_fft)(1);  // handle this special case
+  for (int32 i = 1; i < half_dim; i++) {
+    Real real = (*complex_fft)(i*2), im = (*complex_fft)(i*2 + 1);
+    (*complex_fft)(i) = real*real + im*im;
+  }
+  (*complex_fft)(0) = first_energy;
+  (*complex_fft)(half_dim) = last_energy;  // Will actually never be used, and anyway
+  // if the signal has been bandlimited sensibly this should be zero.
+}
+
+template void ComputePowerSpectrum(VectorBase<float> *complex_fft);
+template void ComputePowerSpectrum(VectorBase<double> *complex_fft);
 
   // no, letting it be non-power-of-two for now.
   // KALDI_ASSERT(dim > 0 && (dim & (dim-1) == 0));  // make sure a power of two.. actually my FFT code
   // does not require this (dan) but this is better in case we use different code [dan].
 
-  // RealFft(waveform, true);  // true == forward (not inverse) FFT; makes no difference here,
-  // as we just want power spectrum.
+template<class Real>
+void PowerSpecToRealCeps(VectorBase<Real> *power_spectrum) {
+  int32 dim = power_spectrum->Dim();
+  int32 half_dim = dim/2;
+  power_spectrum->Range(0, half_dim+1).ApplyLog();
+  power_spectrum->Range(0, half_dim+1).Scale(0.5);  // square root
+  // Now reconstruct the last N/2 - 1 elements of the symmetric spectrum that
+  // correspond to the negative frequencies.
+  /*for (int32 i = 1; i < half_dim; i++)
+    (*power_spectrum)(dim-i) = (*power_spectrum)(i);
+
+    RealFft(power_spectrum, true);  // Doing FFT not IFFT; this is correct.
+    power_spectrum->Scale(1.0/dim); 
+    Real last_cepstrum = (*power_spectrum)(1);
+  for (int32 i = 1; i < half_dim; i++) {
+    Real real = (*power_spectrum)(i*2),
+      im = (*power_spectrum)(i*2 + 1);
+    (*power_spectrum)(i) = real;
+    KALDI_ASSERT(std::abs(im) <= 1e-4 &&
+                 "Real cepstrum not expected to have imaginary value.");
+  }
+  (*power_spectrum)(half_dim) = last_cepstrum;*/
+  Real niq_power = (*power_spectrum)(half_dim);
+  for (int32 i = 1; i < half_dim; i++) {
+      (*power_spectrum)(dim-2*i) = (*power_spectrum)(half_dim - i);
+      (*power_spectrum)(dim-2*i + 1) = 0;
+  }
+  (*power_spectrum)(1) = niq_power;
+  RealFft(power_spectrum, false);  // Doing IFFT.
+  power_spectrum->Scale(1.0/dim);
+}
+
+template void PowerSpecToRealCeps(VectorBase<float> *power_spectrum);
+template void PowerSpecToRealCeps(VectorBase<double> *power_spectrum);
 
-  // now we have in waveform, first half of complex spectrum
-  // it's stored as [real0, realN/2-1, real1, im1, real2, im2, ...]
+
+template<class Real>
```

Comment:
Hi Dan,
We do believe that ultimately this code could be useful for the kaldi community at large, but we also agree that at this point it would not be of direct use. We will try to keep this code in separate files and only in idlak for now. And I'll try to put the code to google standard.

I'll remove these from the pull request, only keep the fix for pitch (as I originally intended).
When I have time I'll try to run the wsj recipe with some of those "exotic" features, and if we get any interesting results, we'll push again at that point!

Thanks for all the hard work.
Blaise